File: src/appSettings.ts
import type { AudioFormat } from "./types";
import type { MinimaxProviderSettings, MinimaxSynthesisOptions } from "./lib/minimaxOptions";
import type { TextFiltersSettings } from "./lib/textFiltersTypes";
import {
  DEFAULT_TIMELINE_VIEW,
  normalizeTimelineViewMode,
  type TimelineViewMode,
} from "./lib/timelineView";

export type { TimelineViewMode };

export type {
  BuiltinFilterToggles,
  CustomTextFilter,
  TextFilterPreset,
  TextFiltersSettings,
  BuiltinFilterOverrides,
} from "./lib/textFiltersTypes";
export {
  DEFAULT_BUILTIN_TOGGLES,
  defaultTextFiltersSettings,
  newCustomFilter,
  newTextFilterPreset,
  resolveActivePreset,
} from "./lib/textFiltersTypes";

export type SaveMode = "manual" | "auto";

export type TtsProviderId = "google" | "voicebox" | "minimax";
export type VoiceboxServerMode = "external" | "bundled" | "disabled";

export const ALL_TTS_PROVIDERS: TtsProviderId[] = ["google", "voicebox", "minimax"];

export interface ApiProfile {
  id: string;
  name: string;
  api_key: string;
}

export interface CursorIntegration {
  enabled: boolean;
  autoplay: boolean;
  max_sentences: number;
  provider: string;
  model: string;
  voice: string;
  style: string | null;
  format: string | null;
  profile_id: string | null;
  language: string | null;
  engine: string | null;
  minimax_speed: number | null;
  minimax_vol: number | null;
  minimax_pitch: number | null;
  minimax_options?: MinimaxSynthesisOptions | null;
  use_summary_markers: boolean;
  dnd_until_ts: number | null;
  last_install_ts: number | null;
}

export function defaultCursorIntegration(): CursorIntegration {
  return {
    enabled: false,
    autoplay: true,
    max_sentences: 10,
    provider: "minimax",
    model: "speech-2.8-hd",
    voice: "Polish_female_1_sample1",
    style: null,
    format: "mp3",
    profile_id: null,
    language: "pl",
    engine: null,
    minimax_speed: 1,
    minimax_vol: 1,
    minimax_pitch: 0,
    use_summary_markers: true,
    dnd_until_ts: null,
    last_install_ts: null,
  };
}

export interface MinimaxPresetVoice {
  voice_id: string;
  display_name: string;
  language: string;
}

export interface MinimaxClonedVoice {
  voice_id: string;
  name: string;
  created_at: number;
  output_vol?: number | null;
}

export interface VoiceProfileSpeaker {
  speaker: string;
  voice: string;
}

/** Zapisany zestaw parametrów TTS (provider, model, głos, styl itd.). */
export interface TtsVoiceProfile {
  id: string;
  name: string;
  provider: string;
  model: string;
  voice: string;
  style: string | null;
  /** Voice Box — id profilu serwera. */
  profile_id: string | null;
  language: string | null;
  engine: string | null;
  /** Voice Box — włącz przepisywanie tekstu w charakterze (personality). */
  personality_enabled?: boolean | null;
  minimax_speed: number | null;
  minimax_vol: number | null;
  minimax_pitch: number | null;
  minimax_options?: MinimaxSynthesisOptions | null;
  multi_speaker: boolean;
  speakers: VoiceProfileSpeaker[];
  /** Jedna linia ostatniego tekstu wygenerowanego tym profilem. */
  last_preview?: string | null;
  last_preview_at?: number | null;
  /** Skrót szybkiej generacji (synchronizowany z quick_hotkeys). */
  shortcut?: string | null;
  shortcut_enabled?: boolean;
}

export interface QuickHotkeyPreset {
  id: string;
  enabled: boolean;
  name: string;
  shortcut: string;
  provider: string;
  model: string;
  voice: string;
  style: string | null;
  profile_id: string | null;
  language: string | null;
  engine: string | null;
  minimax_speed: number | null;
  minimax_vol: number | null;
  minimax_pitch: number | null;
  minimax_options?: MinimaxSynthesisOptions | null;
  load_editor: boolean;
  autoplay: boolean;
  filter_preset_id: string | null;
  format: string | null;
  /** Odwołanie do zapisanego profilu głosu (`voice_profiles`). */
  voice_profile_id?: string | null;
}

export interface QuickHotkeysSettings {
  enabled: boolean;
  presets: QuickHotkeyPreset[];
}

export function defaultQuickHotkeyPreset(name = "Szybki TTS"): QuickHotkeyPreset {
  return {
    id: crypto.randomUUID(),
    enabled: true,
    name,
    shortcut: "Ctrl+Alt+1",
    provider: "google",
    model: "gemini-2.5-flash-preview-tts",
    voice: "Kore",
    style: "Powiedz spokojnie po polsku:",
    profile_id: null,
    language: "pl",
    engine: null,
    minimax_speed: 1,
    minimax_vol: 1,
    minimax_pitch: 0,
    load_editor: false,
    autoplay: true,
    filter_preset_id: null,
    format: null,
    voice_profile_id: null,
  };
}

export function defaultQuickHotkeysSettings(): QuickHotkeysSettings {
  return {
    enabled: false,
    presets: [defaultQuickHotkeyPreset()],
  };
}

/** Slot szybkiej generacji z paska edytora (Gen Ust 1 / 2). */
export interface EditorQuickGenSlot {
  label: string;
  provider: string;
  model: string;
  voice: string;
  style: string | null;
  profile_id: string | null;
  language: string | null;
  engine: string | null;
  minimax_speed: number | null;
  minimax_vol: number | null;
  minimax_pitch: number | null;
  minimax_options?: MinimaxSynthesisOptions | null;
  filter_preset_id: string | null;
  format: string | null;
  voice_profile_id?: string | null;
}

export interface EditorQuickGenSettings {
  slot1: EditorQuickGenSlot;
  slot2: EditorQuickGenSlot;
}

export function defaultEditorQuickGenSlot(label: string): EditorQuickGenSlot {
  return {
    label,
    provider: "google",
    model: "gemini-2.5-flash-preview-tts",
    voice: "Kore",
    style: "Powiedz spokojnie po polsku:",
    profile_id: null,
    language: "pl",
    engine: null,
    minimax_speed: 1,
    minimax_vol: 1,
    minimax_pitch: 0,
    filter_preset_id: null,
    format: null,
    voice_profile_id: null,
  };
}

export function defaultEditorQuickGenSettings(): EditorQuickGenSettings {
  return {
    slot1: defaultEditorQuickGenSlot("Gen Ust 1"),
    slot2: defaultEditorQuickGenSlot("Gen Ust 2"),
  };
}

/** Mapowanie widoku ustawień na payload zapisu (wszystkie pola AppSettings). */
export function appSettingsViewToPayload(view: AppSettingsView): AppSettings {
  return {
    save_mode: view.save_mode,
    save_format: view.save_format,
    temp_path: view.temp_path,
    archive_path: view.archive_path,
    api_profiles: view.api_profiles,
    active_api_id: view.active_api_id,
    cursor_integration: view.cursor_integration,
    max_concurrent_jobs: view.max_concurrent_jobs,
    active_skin_id: view.active_skin_id,
    skin_registry_urls: view.skin_registry_urls ?? [],
    text_filters: view.text_filters,
    minimax_cloned_voices: view.minimax_cloned_voices,
    minimax_synced_voices: view.minimax_synced_voices,
    minimax_voices_synced_at: view.minimax_voices_synced_at ?? null,
    quick_hotkeys: view.quick_hotkeys,
    editor_quick_gen: view.editor_quick_gen ?? defaultEditorQuickGenSettings(),
    voice_profiles: view.voice_profiles ?? [],
    reroute_voice_profile_id: view.reroute_voice_profile_id ?? null,
    quick_setup_completed: view.quick_setup_completed,
    ui_tutorial_completed: view.ui_tutorial_completed,
    enabled_providers: view.enabled_providers,
    minimax_enabled_languages: view.minimax_enabled_languages,
    voicebox_base_url: view.voicebox_base_url ?? null,
    voicebox_server_mode: view.voicebox_server_mode ?? "external",
    minimax_api_key: view.minimax_api_key ?? null,
    minimax_provider_settings: view.minimax_provider_settings,
    temp_history_max: view.temp_history_max ?? DEFAULT_TEMP_HISTORY_MAX,
    quick_history_page_size: view.quick_history_page_size ?? DEFAULT_QUICK_HISTORY_PAGE_SIZE,
    timeline_view: normalizeTimelineViewMode(view.timeline_view),
    safe_mode: view.safe_mode ?? false,
    safe_mode_auto_open_queue: view.safe_mode_auto_open_queue ?? true,
    default_video_template_id: view.default_video_template_id ?? "builtin-whatsapp-karaoke",
    auto_archive_mp4_on_clipboard: view.auto_archive_mp4_on_clipboard ?? true,
  };
}

export interface AppSettings {
  save_mode: SaveMode;
  save_format: AudioFormat;
  temp_path: string | null;
  archive_path: string | null;
  api_profiles: ApiProfile[];
  active_api_id: string | null;
  cursor_integration: CursorIntegration;
  max_concurrent_jobs: number;
  active_skin_id: string;
  skin_registry_urls: string[];
  text_filters?: TextFiltersSettings;
  minimax_cloned_voices?: MinimaxClonedVoice[];
  minimax_synced_voices?: MinimaxPresetVoice[];
  minimax_voices_synced_at?: number | null;
  quick_hotkeys?: QuickHotkeysSettings;
  editor_quick_gen?: EditorQuickGenSettings;
  voice_profiles?: TtsVoiceProfile[];
  /** Gdy ustawione — wszystkie żądania generacji (poza roleplay / skrótami) idą tym profilem. */
  reroute_voice_profile_id?: string | null;
  quick_setup_completed?: boolean;
  /** First-run interactive tutorial (Quick Setup → TTS tour → README summary). */
  ui_tutorial_completed?: boolean;
  enabled_providers?: TtsProviderId[];
  /** Hub codes (`pl`, `en`). Empty = all catalog languages. */
  minimax_enabled_languages?: string[];
  voicebox_base_url?: string | null;
  /** external (default) | bundled (TTS Hub spawns fork) | disabled */
  voicebox_server_mode?: VoiceboxServerMode;
  minimax_api_key?: string | null;
  minimax_provider_settings?: MinimaxProviderSettings;
  /** Max temp history from prior app sessions; current session always kept in full. */
  temp_history_max?: number;
  /** Initial row count (and load-more step) for sidebar "Ostatnie generacje". */
  quick_history_page_size?: number;
  /** Main playback bar waveform style: bars | bars-detailed | line */
  timeline_view?: TimelineViewMode;
  /** Hold new generations for manual approval before synthesis. */
  safe_mode?: boolean;
  /** Expand queue panel and show approval tab when a new pending item arrives. */
  safe_mode_auto_open_queue?: boolean;
  default_video_template_id?: string | null;
  auto_archive_mp4_on_clipboard?: boolean;
}

export { DEFAULT_TIMELINE_VIEW };

export const DEFAULT_MINIMAX_LANGUAGE = "pl";
export const DEFAULT_MINIMAX_VOICE_ID = "Polish_female_1_sample1";

export const MIN_CONCURRENT_JOBS = 1;
export const MAX_CONCURRENT_JOBS = 8;
export const DEFAULT_MAX_CONCURRENT_JOBS = 3;

export const MIN_TEMP_HISTORY_MAX = 10;
export const MAX_TEMP_HISTORY_MAX = 500;
export const DEFAULT_TEMP_HISTORY_MAX = 100;

export const MIN_QUICK_HISTORY_PAGE_SIZE = 5;
export const MAX_QUICK_HISTORY_PAGE_SIZE = 100;
export const DEFAULT_QUICK_HISTORY_PAGE_SIZE = 30;

export interface CursorIntegrationStatus {
  api_ok: boolean;
  hooks_installed: boolean;
  ps1_path: string;
  hooks_json_path: string;
  tts_hub_config_path: string;
  pwsh_available: boolean;
  last_install_ts: number | null;
  last_cursor_at: number | null;
}

export interface McpIntegrationStatus {
  configured: boolean;
  config_path: string | null;
  scope: "global" | "workspace" | null;
}

export interface AppBuildInfo {
  version: string;
  git_hash: string | null;
}

export interface CursorInstallReport {
  copied_ps1: string;
  hooks_json: string;
  hooks_json_backup: string | null;
  tts_hub_config: string;
  ts: number;
}

export interface AppSettingsView extends AppSettings {
  effective_temp_path: string;
  effective_archive_path: string;
  env_api_key_available: boolean;
  env_minimax_api_key_available: boolean;
  effective_voicebox_url: string;
  env_voicebox_url: string;
  effective_minimax_configured: boolean;
}

export function isProviderEnabled(
  enabled: TtsProviderId[] | undefined,
  id: TtsProviderId,
): boolean {
  if (!enabled || enabled.length === 0) return true;
  return enabled.includes(id);
}

export function newApiProfile(name = "Profil API", apiKey = ""): ApiProfile {
  return {
    id: crypto.randomUUID(),
    name,
    api_key: apiKey,
  };
}

File: src-tauri/src/app_settings.rs
use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use std::path::Path;
use uuid::Uuid;

use crate::editor_quick_gen::EditorQuickGenSettings;
use crate::minimax::{
    self, MinimaxClonedVoice, MinimaxPresetVoice, DEFAULT_MINIMAX_LANGUAGE,
    DEFAULT_MINIMAX_VOICE_ID,
};
use crate::quick_hotkeys::QuickHotkeysSettings;
use crate::text_filters::TextFiltersSettings;
use crate::voice_profiles::{normalize_voice_profiles, TtsVoiceProfile};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SaveMode {
    Manual,
    Auto,
}

impl Default for SaveMode {
    fn default() -> Self {
        Self::Manual
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ApiProfile {
    pub id: String,
    pub name: String,
    pub api_key: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CursorIntegration {
    #[serde(default)]
    pub enabled: bool,
    #[serde(default = "default_true")]
    pub autoplay: bool,
    #[serde(default = "default_max_sentences")]
    pub max_sentences: u32,
    #[serde(default = "default_cursor_provider")]
    pub provider: String,
    #[serde(default = "default_cursor_model")]
    pub model: String,
    #[serde(default = "default_cursor_voice")]
    pub voice: String,
    #[serde(default)]
    pub style: Option<String>,
    #[serde(default)]
    pub format: Option<String>,
    #[serde(default)]
    pub profile_id: Option<String>,
    #[serde(default)]
    pub language: Option<String>,
    #[serde(default)]
    pub engine: Option<String>,
    #[serde(default)]
    pub minimax_speed: Option<f32>,
    #[serde(default)]
    pub minimax_vol: Option<f32>,
    #[serde(default)]
    pub minimax_pitch: Option<i32>,
    #[serde(default = "default_true")]
    pub use_summary_markers: bool,
    #[serde(default)]
    pub dnd_until_ts: Option<i64>,
    #[serde(default)]
    pub last_install_ts: Option<i64>,
}

fn default_true() -> bool {
    true
}
fn default_max_sentences() -> u32 {
    10
}
fn default_cursor_provider() -> String {
    PROVIDER_MINIMAX.to_string()
}
fn default_cursor_model() -> String {
    "speech-2.8-hd".to_string()
}
fn default_cursor_voice() -> String {
    DEFAULT_MINIMAX_VOICE_ID.to_string()
}

impl CursorIntegration {
    pub fn normalize(&mut self) {
        self.provider = self.provider.trim().to_lowercase();
        if !ALL_PROVIDERS.contains(&self.provider.as_str()) {
            self.provider = default_cursor_provider();
        }
        self.model = self.model.trim().to_string();
        self.voice = self.voice.trim().to_string();
        if self.model.is_empty() {
            self.model = match self.provider.as_str() {
                PROVIDER_MINIMAX => "speech-2.8-hd".to_string(),
                PROVIDER_VOICEBOX => "voicebox:chatterbox".to_string(),
                _ => "gemini-2.5-flash-preview-tts".to_string(),
            };
        }
        if self.voice.is_empty() {
            self.voice = match self.provider.as_str() {
                PROVIDER_MINIMAX => minimax::MinimaxClient::default_voice_for_language(
                    self.language.as_deref().unwrap_or(DEFAULT_MINIMAX_LANGUAGE),
                )
                .to_string(),
                PROVIDER_VOICEBOX => String::new(),
                _ => "Kore".to_string(),
            };
        }
        if let Some(fmt) = self.format.as_mut() {
            *fmt = fmt.trim().to_lowercase();
            if !matches!(fmt.as_str(), "wav" | "mp3" | "ogg") {
                *fmt = String::new();
            }
            if fmt.is_empty() {
                self.format = None;
            }
        }
        if self.format.is_none() {
            self.format = Some(match self.provider.as_str() {
                PROVIDER_MINIMAX => "mp3".to_string(),
                _ => "wav".to_string(),
            });
        }
        if let Some(speed) = self.minimax_speed {
            self.minimax_speed = Some(speed.clamp(0.5, 2.0));
        }
        if let Some(vol) = self.minimax_vol {
            self.minimax_vol = Some(vol.clamp(0.0, 10.0));
        }
        if let Some(pitch) = self.minimax_pitch {
            self.minimax_pitch = Some(pitch.clamp(-12, 12));
        }
        if self.provider != PROVIDER_MINIMAX {
            self.minimax_speed = None;
            self.minimax_vol = None;
            self.minimax_pitch = None;
        } else if self.minimax_speed.is_none() {
            self.minimax_speed = Some(1.0);
            self.minimax_vol = Some(1.0);
            self.minimax_pitch = Some(0);
        }
        if self.provider == PROVIDER_VOICEBOX {
            if self.language.is_none() {
                self.language = Some(DEFAULT_MINIMAX_LANGUAGE.to_string());
            }
            self.engine = self.engine.take().filter(|e| !e.trim().is_empty());
        } else if self.provider == PROVIDER_MINIMAX {
            self.profile_id = None;
            self.engine = None;
            let lang = self
                .language
                .take()
                .map(|l| l.trim().to_ascii_lowercase())
                .filter(|l| minimax::is_known_language_code(l));
            self.language = Some(lang.unwrap_or_else(|| DEFAULT_MINIMAX_LANGUAGE.to_string()));
        } else {
            self.profile_id = None;
            self.language = None;
            self.engine = None;
        }
    }
}

impl Default for CursorIntegration {
    fn default() -> Self {
        Self {
            enabled: false,
            autoplay: true,
            max_sentences: 10,
            provider: default_cursor_provider(),
            model: default_cursor_model(),
            voice: default_cursor_voice(),
            style: None,
            format: Some("mp3".to_string()),
            profile_id: None,
            language: Some(DEFAULT_MINIMAX_LANGUAGE.to_string()),
            engine: None,
            minimax_speed: Some(1.0),
            minimax_vol: Some(1.0),
            minimax_pitch: Some(0),
            use_summary_markers: true,
            dnd_until_ts: None,
            last_install_ts: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AppSettings {
    #[serde(default)]
    pub save_mode: SaveMode,
    #[serde(default = "default_save_format")]
    pub save_format: String,
    pub temp_path: Option<String>,
    pub archive_path: Option<String>,
    #[serde(default)]
    pub api_profiles: Vec<ApiProfile>,
    pub active_api_id: Option<String>,
    #[serde(default)]
    pub cursor_integration: CursorIntegration,
    #[serde(default = "default_max_concurrent_jobs")]
    pub max_concurrent_jobs: u32,
    #[serde(default = "default_active_skin_id")]
    pub active_skin_id: String,
    #[serde(default)]
    pub skin_registry_urls: Vec<String>,
    #[serde(default)]
    pub text_filters: TextFiltersSettings,
    #[serde(default)]
    pub minimax_cloned_voices: Vec<MinimaxClonedVoice>,
    /// System voices from last `get_voice` API sync; empty = built-in catalog.
    #[serde(default)]
    pub minimax_synced_voices: Vec<MinimaxPresetVoice>,
    #[serde(default)]
    pub minimax_voices_synced_at: Option<i64>,
    #[serde(default)]
    pub quick_hotkeys: QuickHotkeysSettings,
    #[serde(default)]
    pub editor_quick_gen: EditorQuickGenSettings,
    #[serde(default)]
    pub voice_profiles: Vec<TtsVoiceProfile>,
    /// When set, incoming generation requests (except roleplay / quick hotkeys)
    /// are forced through this saved voice profile regardless of caller params.
    #[serde(default)]
    pub reroute_voice_profile_id: Option<String>,
    #[serde(default)]
    pub quick_setup_completed: bool,
    #[serde(default)]
    pub enabled_providers: Vec<String>,
    #[serde(default = "default_minimax_enabled_languages")]
    pub minimax_enabled_languages: Vec<String>,
    pub voicebox_base_url: Option<String>,
    pub minimax_api_key: Option<String>,
    /// Max non-archived temp history rows from prior app sessions (current session always kept).
    #[serde(default = "default_temp_history_max")]
    pub temp_history_max: u32,
    /// Main playback bar waveform: bars | bars-detailed | line
    #[serde(default = "default_timeline_view")]
    pub timeline_view: String,
    /// When true, new generations are held for user approval before synthesis.
    #[serde(default)]
    pub safe_mode: bool,
    /// Expand the queue panel and switch to approval tab when a pending item arrives.
    #[serde(default = "default_safe_mode_auto_open_queue")]
    pub safe_mode_auto_open_queue: bool,
}

fn default_safe_mode_auto_open_queue() -> bool {
    true
}

fn default_minimax_enabled_languages() -> Vec<String> {
    vec![DEFAULT_MINIMAX_LANGUAGE.to_string()]
}

pub const PROVIDER_GOOGLE: &str = "google";
pub const PROVIDER_VOICEBOX: &str = "voicebox";
pub const PROVIDER_MINIMAX: &str = "minimax";

const ALL_PROVIDERS: &[&str] = &[PROVIDER_GOOGLE, PROVIDER_VOICEBOX, PROVIDER_MINIMAX];

fn default_active_skin_id() -> String {
    "vibelife".to_string()
}

fn default_save_format() -> String {
    "wav".to_string()
}

pub const MIN_CONCURRENT_JOBS: u32 = 1;
pub const MAX_CONCURRENT_JOBS_CAP: u32 = 8;

fn default_max_concurrent_jobs() -> u32 {
    3
}

pub const MIN_TEMP_HISTORY_MAX: u32 = 10;
pub const MAX_TEMP_HISTORY_MAX: u32 = 500;

fn default_temp_history_max() -> u32 {
    100
}

fn default_timeline_view() -> String {
    "bars".to_string()
}

const TIMELINE_VIEWS: &[&str] = &["bars", "bars-detailed", "line"];

impl Default for AppSettings {
    fn default() -> Self {
        Self {
            save_mode: SaveMode::Manual,
            save_format: default_save_format(),
            temp_path: None,
            archive_path: None,
            api_profiles: Vec::new(),
            active_api_id: None,
            cursor_integration: CursorIntegration::default(),
            max_concurrent_jobs: default_max_concurrent_jobs(),
            active_skin_id: default_active_skin_id(),
            skin_registry_urls: Vec::new(),
            text_filters: TextFiltersSettings::default(),
            minimax_cloned_voices: Vec::new(),
            minimax_synced_voices: Vec::new(),
            minimax_voices_synced_at: None,
            quick_hotkeys: QuickHotkeysSettings::default(),
            editor_quick_gen: EditorQuickGenSettings::default(),
            voice_profiles: Vec::new(),
            reroute_voice_profile_id: None,
            quick_setup_completed: false,
            enabled_providers: Vec::new(),
            minimax_enabled_languages: default_minimax_enabled_languages(),
            voicebox_base_url: None,
            minimax_api_key: None,
            temp_history_max: default_temp_history_max(),
            timeline_view: default_timeline_view(),
            safe_mode: false,
            safe_mode_auto_open_queue: default_safe_mode_auto_open_queue(),
        }
    }
}

/// Strips UTF-8 BOM and surrounding whitespace (PowerShell / some editors add BOM).
fn prepare_settings_json(raw: &str) -> &str {
    raw.strip_prefix('\u{feff}').unwrap_or(raw).trim()
}

impl AppSettings {
    pub fn load(path: &Path) -> Result<Self> {
        if !path.exists() {
            return Ok(Self::default());
        }
        let raw =
            std::fs::read_to_string(path).with_context(|| format!("read {}", path.display()))?;
        let json = prepare_settings_json(&raw);
        if json.is_empty() {
            return Ok(Self::default());
        }

        match serde_json::from_str::<Self>(json) {
            Ok(mut settings) => {
                settings.normalize();
                Ok(settings)
            }
            Err(e) => {
                let backup = path.with_extension("json.bak");
                if std::fs::copy(path, &backup).is_ok() {
                    eprintln!(
                        "settings.json: parse failed ({e}); backup at {}",
                        backup.display()
                    );
                } else {
                    eprintln!("settings.json: parse failed ({e})");
                }
                Ok(Self::default())
            }
        }
    }

    pub fn save(&self, path: &Path) -> Result<()> {
        if let Some(parent) = path.parent() {
            std::fs::create_dir_all(parent)?;
        }
        let mut copy = self.clone();
        copy.normalize();
        let raw = serde_json::to_string_pretty(&copy)?;
        std::fs::write(path, raw)?;
        Ok(())
    }

    pub fn active_google_key(&self, env_key: &str) -> String {
        if let Some(id) = &self.active_api_id {
            if let Some(profile) = self.api_profiles.iter().find(|p| p.id == *id) {
                let key = profile.api_key.trim();
                if !key.is_empty() {
                    return key.to_string();
                }
            }
        }
        env_key.to_string()
    }

    pub fn effective_voicebox_url(&self, env_url: &str) -> String {
        self.voicebox_base_url
            .as_ref()
            .map(|s| s.trim())
            .filter(|s| !s.is_empty())
            .map(|s| s.trim_end_matches('/').to_string())
            .unwrap_or_else(|| {
                let e = env_url.trim();
                if e.is_empty() {
                    "http://127.0.0.1:17493".to_string()
                } else {
                    e.trim_end_matches('/').to_string()
                }
            })
    }

    pub fn effective_minimax_key(&self, env_key: &str) -> String {
        self.minimax_api_key
            .as_ref()
            .map(|s| s.trim())
            .filter(|s| !s.is_empty())
            .map(|s| s.to_string())
            .unwrap_or_else(|| env_key.trim().to_string())
    }

    pub fn is_provider_enabled(&self, id: &str) -> bool {
        if self.enabled_providers.is_empty() {
            return true;
        }
        self.enabled_providers.iter().any(|p| p == id)
    }

    pub fn normalize(&mut self) {
        self.save_format = self.save_format.trim().to_lowercase();
        if !matches!(self.save_format.as_str(), "wav" | "mp3" | "ogg") {
            self.save_format = default_save_format();
        }
        self.temp_path = normalize_optional_path(self.temp_path.take());
        self.archive_path = normalize_optional_path(self.archive_path.take());
        for profile in &mut self.api_profiles {
            if profile.id.trim().is_empty() {
                profile.id = Uuid::new_v4().to_string();
            }
            profile.name = profile.name.trim().to_string();
            if profile.name.is_empty() {
                profile.name = "Profil API".to_string();
            }
        }
        if let Some(id) = &self.active_api_id {
            if !self.api_profiles.iter().any(|p| p.id == *id) {
                self.active_api_id = None;
            }
        }
        if self.max_concurrent_jobs < MIN_CONCURRENT_JOBS {
            self.max_concurrent_jobs = MIN_CONCURRENT_JOBS;
        } else if self.max_concurrent_jobs > MAX_CONCURRENT_JOBS_CAP {
            self.max_concurrent_jobs = MAX_CONCURRENT_JOBS_CAP;
        }
        if self.temp_history_max < MIN_TEMP_HISTORY_MAX {
            self.temp_history_max = MIN_TEMP_HISTORY_MAX;
        } else if self.temp_history_max > MAX_TEMP_HISTORY_MAX {
            self.temp_history_max = MAX_TEMP_HISTORY_MAX;
        }
        self.active_skin_id = self.active_skin_id.trim().to_string();
        if self.active_skin_id.is_empty() {
            self.active_skin_id = default_active_skin_id();
        }
        self.skin_registry_urls = self
            .skin_registry_urls
            .iter()
            .map(|u| u.trim().to_string())
            .filter(|u| !u.is_empty())
            .collect();
        self.text_filters.normalize();
        self.cursor_integration.normalize();
        self.quick_hotkeys.normalize();
        self.editor_quick_gen.normalize();
        normalize_voice_profiles(&mut self.voice_profiles);
        if let Some(id) = self.reroute_voice_profile_id.take() {
            let trimmed = id.trim().to_string();
            if trimmed.is_empty() || !self.voice_profiles.iter().any(|p| p.id == trimmed) {
                self.reroute_voice_profile_id = None;
            } else {
                self.reroute_voice_profile_id = Some(trimmed);
            }
        }
        self.voicebox_base_url = normalize_optional_path(self.voicebox_base_url.take());
        self.minimax_api_key = self
            .minimax_api_key
            .take()
            .map(|s| s.trim().to_string())
            .filter(|s| !s.is_empty());
        let mut seen = std::collections::HashSet::new();
        self.enabled_providers = self
            .enabled_providers
            .iter()
            .map(|p| p.trim().to_lowercase())
            .filter(|p| ALL_PROVIDERS.contains(&p.as_str()))
            .filter(|p| seen.insert(p.clone()))
            .collect();
        let mut lang_seen = std::collections::HashSet::new();
        self.minimax_enabled_languages = self
            .minimax_enabled_languages
            .iter()
            .map(|c| c.trim().to_ascii_lowercase())
            .filter(|c| minimax::is_known_language_code(c))
            .filter(|c| lang_seen.insert(c.clone()))
            .collect();
        self.timeline_view = self.timeline_view.trim().to_ascii_lowercase();
        if !TIMELINE_VIEWS.contains(&self.timeline_view.as_str()) {
            self.timeline_view = default_timeline_view();
        }
        for v in &mut self.minimax_cloned_voices {
            v.normalize_output_vol();
            if v.voice_id.eq_ignore_ascii_case("robert_maklowicz") && v.output_vol.is_none() {
                v.output_vol = Some(2.0);
            }
        }
    }

    pub fn effective_minimax_enabled_languages(&self) -> Vec<String> {
        minimax::effective_enabled_language_codes(&self.minimax_enabled_languages)
    }
}

fn normalize_optional_path(value: Option<String>) -> Option<String> {
    value
        .map(|s| s.trim().to_string())
        .filter(|s| !s.is_empty())
}

pub fn new_api_profile(name: impl Into<String>, api_key: impl Into<String>) -> ApiProfile {
    ApiProfile {
        id: Uuid::new_v4().to_string(),
        name: name.into(),
        api_key: api_key.into(),
    }
}
